### ISSUE TITLE
Predefined schemas marked as invalid

### ISSUE THREAD

**Issue by schmkr** [mulesoft/api-workbench#53]
Hi there,

Love what you've created with this project. We use RAML to define our API and generate interfaces from the (json) schemas in various languages (C#, Java & TypeScript) as well as multiple validation packages (`ramlcop`, `ramlev`). We have defined those schemas in external json files and include those in the top of our RAML (0.8) document like this:

```
schemas:
- common/definitions.json: !include schemas/common/definitions.json
- version/getVersion.json: !include schemas/version/getVersion.json
```

and in our resource definitions, we refer to those schemas:

```
/version:
  displayName: Version Information
  description: API information
  get:
    responses:
      200:
        body:
          application/json:
            schema: version/getVersion.json
```

We do it this way, because of [an issue with raml-for-jax-rs](https://github.com/mulesoft/raml-for-jax-rs/issues/125). 

This works, and everything we use to parse the RAML validated the referred schemas. Api Workbench however, does recognize the defined schemas (they end up in the autocomplete and the select option of the details pane), but it does result in a validation error:

`Syntax error: Expected "|" or end of input but "/" found.`

I already tried surrounding the schemas in quotes, but that doesn't make any difference.

**Comment by ddenisenko** [comment]
@schmkr Thank you for reporting that. The root of the issue is that your schema names contain slash character.

If you change your code to:

```RAML
schemas:
- common-definitions.json: !include schemas/common/definitions.json
- version-getVersion.json: !include schemas/version/getVersion.json
```

```RAML
/version:
  displayName: Version Information
  description: API information
  get:
    responses:
      200:
        body:
          application/json:
            schema: version-getVersion.json
```

or something like that, it will work fine.

I agree that validation in the case should be more meaningful and should blame that slashes as soon as they appear.

**Comment by usarid** [comment]
But why can't these names/keys contain slashes, perhaps as long as they're quoted?






On Thu, Nov 12, 2015 at 12:50 PM -0800, "Denis Denisenko" <notifications@github.com> wrote:












@schmkr Thank you for reporting that. The root of the issue is that your schema names contain slash character.



If you change your code to:

schemas:
- common-definitions.json: !include schemas/common/definitions.json
- version-getVersion.json: !include schemas/version/getVersion.json

/version:
  displayName: Version Information
  description: API information
  get:
    responses:
      200:
        body:
          application/json:
            schema: version-getVersion.json



or something like that, it will work fine.



I agree that validation in the case should be more meaningful and should blame that slashes as soon as they appear.

**Comment by ddenisenko** [comment]
Well, because we want to stay close in naming rules with the languages we're going to generate code at. In most of them using slashes in variables names, type names etc is prohibited.

**Comment by usarid** [comment]
Not sure that's consistent with anything we've done before, which is
important: I believe keys have been allowed to be anything valid in YAML,
so we'd have to think a bunch before deciding we want to make an exception
here. Besides, why would we regard these as potential variables? They're
just internal names.

**Comment by schmkr** [comment]
Thanks @usarid. I agree.

@ddenisenko As said, we are using various parsers and every one of them does not have an issue with the names of the schemas and afaik they are according to spec. We went down the route of using these names for the schemas because of a bug in the [raml-for-jax-rs](https://github.com/mulesoft/raml-for-jax-rs/issues/125) parser.

**Comment by schmkr** [comment]
@usarid By the way, I've already tried using quotes, but this made no difference.

**Comment by sichvoge** [comment]
IMHO, maybe we should fix the bug in the raml-for-jsx-rs parser. My reason being is that I would follow some standards around naming convention for keys that are more general. I agree that in your case that might work with the current parser, but we don't know what others will expect. Uri said that something like that is allowed in YAML, but not everything that is allowed is best practice. That is why we have them defined. 

Therefore, my personal take on it is to restrict on `/`. Not saying that I am not open to anything else ;)

**Comment by ddenisenko** [comment]
>Besides, why would we regard these as potential variables? They're
just internal names.

Example from the recent Invoices API:

```RAML
types:
  ...
  InternalInvoice:
    type: Invoice
    properties:
      internalCode: number
  ...
```

translates to:

```java
public class InternalInvoice {
...
}
```

So the type name from RAML directly translates to a Java class name.

Of course we can add slash replacement in all client gens and sever gens, but do we want to?

**Comment by usarid** [comment]
Yes, I think that every client and server generation language will have its own rules on capitalization, escaping, etc. RAML itself should set some reasonable limits on keys beyond what YAML sets (they need to be scalars, their string representation must be between 1 and N characters long where N is...), and that's all.






On Sat, Nov 14, 2015 at 5:10 PM -0800, "Denis Denisenko" <notifications@github.com> wrote:













Besides, why would we regard these as potential variables? They're

just internal names.




Example from the recent Invoices API:

types:
  ...
  InternalInvoice:
    type: Invoice
    properties:
      internalCode: number
  ...



translates to:

public class InternalInvoice {
...
}



So the type name from RAML directly translates to a Java class name.



Of course we can add slash replacement in all client gens and sever gens, but do we want to?

**Comment by schmkr** [comment]
Any update on this? The issue seems to be only in the validator. The schemas get succesfully parsed in the autocomplete list: 

    schemas:
      - contacts/definitions.json: !include schemas/contacts/definitions.json
      - contacts/entityContactRequest.json: !include schemas/contacts/entityContactRequest.json
      - contacts/entityContactResponse.json: !include schemas/contacts/entityContactResponse.json
      - contacts/getContacts200.json: !include schemas/contacts/getContacts200.json
      - contacts/postContactPictureResponse.json: !include schemas/contacts/postContactPictureResponse.json
      - conversations/entityConversationResponse.json: !include schemas/conversations/entityConversationResponse.json
      - conversations/entityParticipant.json: !include schemas/conversations/entityParticipant.json
      - conversations/getConversations200.json: !include schemas/conversations/getConversations200.json

![autocomplete list](https://s3.amazonaws.com/f.cl.ly/items/1d2F0b2l2c120l2c3S30/Screen%20Shot%202015-12-23%20at%2015.08.35.png)!

Only the validator does not validate the name of the schema, every project we use to generate classes from the RAML and its schemas does work, also [raml-for-jax-rs](https://github.com/mulesoft/raml-for-jax-rs).

Whenever users would run into invalid generated class or interface names, it would be up to the user to change their schema names, right?
